src/python/netvlad/netvlad_extractor.py

```diff
@@ -0,0 +1,217 @@
+# The parsing of the .mat file is according to aforementioned project.
+
+import torch
+from torch import nn
+from torchvision.transforms import ToTensor
+import torchvision.models as models
+from torch.utils.data import Dataset
+
+from PIL import Image
+
+from scipy.io import loadmat
+import numpy as np
+
+from pathlib import Path
+import argparse
+
+
+class ImageDataset(Dataset):
+    def __init__(self, img_dir, transform=None):
+        self.img_names = []
+        self.img_names.extend(img_dir.glob("*.png"))
+        self.img_names.extend(img_dir.glob("*.jpg"))
+        self.img_names.sort()
+        self.transform = transform
+
+    def __len__(self):
+        return len(self.img_names)
+
+    def __getitem__(self, idx):
+        assert idx >= 0 and idx < len(self.img_names)
+        image = Image.open(self.img_names[idx].as_posix())
+        image = image.resize((224, 224))
+        if self.transform:
+            image = self.transform(image)
+        # returns an image in the proper format
+        return image
+
+
+class VLADLayer(nn.Module):
+    def __init__(
+        self,
+        inputDim=512,
+        numberOfClusters=64,
+    ):
+        super().__init__()
+        # Sort of clustering.
+        # Assigning a feature and clusters with softmax assignment
+        # Initialize the needed layers.
+        self.clusterAssociations = torch.nn.Conv1d(
+            inputDim, numberOfClusters, 1, stride=1
+        )
+        centers = torch.nn.parameter.Parameter(
+            torch.empty([inputDim, numberOfClusters])
+        )
+        self.register_parameter("centers", centers)
+
+    def forward(self, featureVolume):
+        # featureVolume of size (1, D, N)
+        batch = featureVolume.size(0)
+        # Size (1, K, N)
+        assignmentScores = self.clusterAssociations(featureVolume)
+        softAssignments = nn.functional.softmax(assignmentScores, dim=1)
+
+        # Reshape from (1, D, N) to (1, D, 1, N)
+        featureTensor = featureVolume.unsqueeze(2)
+        # Reshape from (D,K) to size (1, D, K, 1)
+        centersTensor = self.centers.unsqueeze(0).unsqueeze(-1)
+        # VLAD core (notation from paper Eq 4).
+        descriptors = (
+            softAssignments.unsqueeze(1) * (featureTensor - centersTensor)
+        ).sum(dim=-1)
+
+        # Intra normalization
+        # Descriptors size (1, D, K)
+        descriptors = nn.functional.normalize(descriptors, dim=1)
+
+        # L2 normalization
+        descriptor = descriptors.view(batch, -1)
+        descriptor = nn.functional.normalize(descriptor, dim=1)
+        return descriptor
+
+
+class NetVLAD(nn.Module):
+    def __init__(self, pathToMat):
+        super(NetVLAD, self).__init__()
+        # Define overall network structure.
+        vgg16Features = list(models.vgg16().children())[0]
+        # Remove last ReLU + MaxPool2d.
+        self.backbone = nn.Sequential(*list(vgg16Features.children())[:-2])
+        inputDim = 512
```

Comment:
Do we maybe want to configure this externally? Maybe from a config file? Or do you expect this to never change? Same for all other magic constants.

Reply:
Yes, this only works because the weights I am reading in and stored for selected dimensions.